coil-gif/src/nonAndroidMain/kotlin/coil3/gif/AnimatedSkiaImage.kt

```diff
@@ -0,0 +1,233 @@
+
+        if (isAnimationComplete) {
+            // The animation is complete, freeze on the last frame.
+            canvas.drawFrame(
+                frameIndex = codec.framesInfo.lastIndex,
+            )
+
+            if (!hasNotifiedAnimationEnd) {
+                onAnimationEnd?.invoke()
+                hasNotifiedAnimationEnd = true
+            }
+
+            return
+        }
+
+        val frameIndexToDraw = getFrameIndexToDraw(
+            frameDurationsMs = frameDurationsMs,
+            singleIterationDurationMs = singleIterationDurationMs,
+            totalElapsedTimeMs = totalElapsedTimeMs,
+        )
+
+        canvas.drawFrame(frameIndexToDraw)
+
+        if (!isAnimationComplete) {
+            invalidateTick++
+        }
+    }
+
+    /**
+     * Returns the index of the frame to draw based on the current time.
+     *
+     * If the animation is complete, the index of the last frame is returned.
+     *
+     * @param frameDurationsMs The list of frame durations in milliseconds.
+     */
+    private fun getFrameIndexToDraw(
+        frameDurationsMs: List<Int>,
+        singleIterationDurationMs: Int,
+        totalElapsedTimeMs: Long,
+    ): Int {
+        val currentIterationElapsedTimeMs = totalElapsedTimeMs % singleIterationDurationMs
+        return getFrameIndexToDrawWithinIteration(
+            frameDurationsMs = frameDurationsMs,
+            elapsedTimeMs = currentIterationElapsedTimeMs,
+        )
+    }
+
+    /**
+     * Returns the index of the frame to draw within the current iteration.
+     *
+     * @param frameDurationsMs The list of frame durations in milliseconds.
+     * @param elapsedTimeMs The elapsed time in milliseconds since the start of the iteration.
+     */
+    private fun getFrameIndexToDrawWithinIteration(
+        frameDurationsMs: List<Int>,
+        elapsedTimeMs: Long,
+    ): Int {
+        var accumulatedDuration = 0
+
+        for ((index, frameDuration) in frameDurationsMs.withIndex()) {
+            if (accumulatedDuration > elapsedTimeMs) {
+                return (index - 1).coerceAtLeast(0)
+            }
+
+            accumulatedDuration += frameDuration
+        }
+
+        return frameDurationsMs.lastIndex
+    }
+
+    private fun decodeFrame(frameIndex: Int): ByteArray {
+        if (tempBitmap.isClosed) {
+            throw IllegalStateException("Cannot decode frame: the bitmap is closed.")
+        }
+
+        codec.readPixels(tempBitmap, frameIndex)
+        return tempBitmap.readPixels(
+            dstInfo = codec.imageInfo,
+            dstRowBytes = codec.imageInfo.minRowBytes,
+        )!!
+    }
+
+    private fun Canvas.drawFrame(frameIndex: Int) {
+        animatedTransformation?.transform(this)
+
+        frames[frameIndex]?.let { frame ->
+            drawImage(
+                image = org.jetbrains.skia.Image.makeRaster(
+                    imageInfo = codec.imageInfo,
+                    bytes = frame,
```

Comment:
@revonateB0T Agreed, but let's tackle this as a follow up.

For buffer size we need to consume/buffer the `ImageSource` entirely inside `Decode.decode`. This is to prevent leaking underlying references (mostly to Coil's disk cache). Would that still work with decode-render-swap?

Reply:
@revonateB0T Ah I see what you mean. I was talking about copying the undecoded data in `Decode.decode` so we can read it outside of that scope (similar to what we do in `AnimatedImageDecoder`. How do we solve this in a common way across platforms?

Reply:
@revonateB0T We can't use a file backed source as we don't know when the animated GIF is no longer used and we can't expect users to manually call a `close` method.

Reply:
Yep, we can't rely on that behavior on all supported platforms. Also Skiko [doesn't support streaming](https://youtrack.jetbrains.com/issue/SKIKO-741) - all bytes need to be in memory.